contracts/cw-ics721-bridge/src/contract.rs

```diff
@@ -0,0 +1,544 @@
+                    name: class_id.to_string(),
+                    symbol: class_id.to_string(), // TODO: What should we put here?
+                    minter: env.contract.address.to_string(),
+                })?,
+                funds: vec![],
+                // Attempting to fit the class ID in the label field
+                // can make this field too long which causes weird
+                // data errors in the SDK.
+                label: "ICS771 backing cw721".to_string(),
+            },
+            INSTANTIATE_CW721_REPLY_ID,
+        );
+        vec![message]
+    };
+
+    // Mint the requested tokens. Submessages and their replies are
+    // always executed before regular messages [1], so we can sleep
+    // nicely knowing this won't happen until the cw721 contract has
+    // been instantiated.
+    //
+    // [1] https://github.com/CosmWasm/cosmwasm/blob/main/SEMANTICS.md#order-and-rollback
+    let mint_message = WasmMsg::Execute {
+        contract_addr: env.contract.address.into_string(),
+        msg: to_binary(&ExecuteMsg::Mint {
+            class_id,
+            token_ids,
+            token_uris,
+            receiver,
+        })?,
+        funds: vec![],
+    };
+
+    Ok(Response::default()
+        .add_attribute("method", "do_instantiate_and_mint")
+        .add_submessages(submessages)
+        .add_message(mint_message))
+}
+
+fn execute_receive_nft(
+    deps: DepsMut,
+    info: MessageInfo,
+    token_id: String,
+    sender: String,
+    msg: Binary,
+) -> Result<Response, ContractError> {
+    let sender = deps.api.addr_validate(&sender)?;
+    let msg: IbcAwayMsg = from_binary(&msg)?;
+
+    // When we receive a NFT over IBC we want to avoid parsing of the
+    // classId field at all costs. String parsing has interesting time
+    // complexity issues and is error prone.
+    //
+    // We would normally need to parse the classId field when we
+    // receive a NFT from another chain, and the top of the classId
+    // stack is our own chain. In this case, it is possible that upon
+    // that NFT arriving at this chain it will be unpacked back into a
+    // native NFT. To determine this we could either:
+    //
+    // (1) Check if the classId has no '/' characters. Or,
+    // (2) Store each NFT we receive in our maps (giving it a classID
+    //     of its own address) and do the same lookup / transfer
+    //     regardless of if the NFT is becoming a native one.
+    //
+    // To avoid special logic when receiving packets, we go the save
+    // route. We can change this back later if we'd like.
+    if !NFT_CONTRACT_TO_CLASS_ID.has(deps.storage, info.sender.clone()) {
+        // This is a new NFT so we give it a "classID". If the map
+        // does have the key, it means that either (a) we have seen
+        // the contract before, or (b) this is a NFT that we received
+        // over IBC and made a new contract for.
+        NFT_CONTRACT_TO_CLASS_ID.save(
```

Comment:
Need to circle back on this part of the code: 
"recall that, one bridge can send out nfts on multiple channels
so you have bad kids 1 went out on channel 0 to juno and bad kids 2 went out on channel 1 to omni
so you want to prepend channel0/badkids1 and channel1/badkids2
then when you receive back on those channels you know its a valid source"